File: app/src/main/java/jp/juggler/subwaytooter/api/DuplicateMap.java
package jp.juggler.subwaytooter.api;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;

import jp.juggler.subwaytooter.api.entity.TootAccount;
import jp.juggler.subwaytooter.api.entity.TootNotification;
import jp.juggler.subwaytooter.api.entity.TootReport;
import jp.juggler.subwaytooter.api.entity.TootStatus;
import jp.juggler.subwaytooter.api_tootsearch.entity.TSToot;

public class DuplicateMap {
	
	private final HashSet< Long > set_status_id = new HashSet<>();
	private final HashSet< Long > set_notification_id = new HashSet<>();
	private final HashSet< Long > set_report_id = new HashSet<>();
	private final HashSet< Long > set_account_id = new HashSet<>();
	private final HashSet< String > set_status_uri = new HashSet<>();
	
	public void clear(){
		set_status_id.clear();
		set_notification_id.clear();
		set_report_id.clear();
		set_account_id.clear();
		set_status_uri.clear();
	}
	
	private boolean isDuplicate( Object o ){

		if( o instanceof TSToot ){
			TSToot s = (TSToot)o;
			if( set_status_uri.contains( s.uri ) ) return true;
			set_status_uri.add( s.uri  );
			return false;
		}
		
		if( o instanceof TootStatus ){
			if( set_status_id.contains( ( (TootStatus) o ).id ) ) return true;
			set_status_id.add( ( (TootStatus) o ).id );

		}else if( o instanceof TootNotification ){
			if( set_notification_id.contains( ( (TootNotification) o ).id ) ) return true;
			set_notification_id.add( ( (TootNotification) o ).id );

		}else if( o instanceof TootReport ){
			if( set_report_id.contains( ( (TootReport) o ).id ) )  return true;
			set_report_id.add( ( (TootReport) o ).id );

		}else if( o instanceof TootAccount ){
			if( set_account_id.contains( ( (TootAccount) o ).id ) ) return true;
			set_account_id.add( ( (TootAccount) o ).id );
		}

		return false;
	}
	
	public ArrayList<Object> filterDuplicate( Collection< Object > src ){
		ArrayList< Object > list_new = new ArrayList<>();
		for( Object o : src ){
			if( isDuplicate( o )) continue;
			list_new.add( o );
		}

		return list_new;
	}
	
}

File: app/src/main/java/jp/juggler/subwaytooter/api/DuplicateMap.kt
package jp.juggler.subwaytooter.api

import jp.juggler.subwaytooter.api.entity.*
import jp.juggler.util.LogCategory
import jp.juggler.util.cast
import java.util.*

class DuplicateMap {

    companion object {
        val log = LogCategory("DuplicateMap")
    }

    private val idSet = HashSet<EntityId>()
    private val uriSet = HashSet<String>()

    fun clear() {
        idSet.clear()
        uriSet.clear()
    }

    fun isDuplicate(o: TimelineItem): Boolean {

        if (o is TootStatus) {
            val uri = o.uri
            val url = o.url
            when {
                uri.isNotEmpty() -> {
                    if (uriSet.contains(uri)) return true
                    uriSet.add(uri)
                }

                url?.isNotEmpty() == true -> {
                    // URIとURLで同じマップを使いまわすが、害はないと思う…
                    if (uriSet.contains(url)) return true
                    uriSet.add(url)
                }
            }
        }

        when (o) {
            is TootReport,
            is TootStatus,
            is TootAccountRef,
            is TootNotification,
            -> {
                val id = o.getOrderId()
                if (id.notDefaultOrConfirming) {
                    if (idSet.contains(id)) return true
                    idSet.add(id)
                }
            }
        }

        return false
    }

    fun filterDuplicate(src: Collection<TimelineItem>?): ArrayList<TimelineItem> {
        val listNew = ArrayList<TimelineItem>()
        if (src != null) {
            for (o in src) {
                if (isDuplicate(o)) {
                    log.d("filterDuplicate: filter orderId ${o.getOrderId()}")
                    continue
                }
                listNew.add(o)
            }
        }
        return listNew
    }

    private fun isDuplicateByCreatedAt(o: TimelineItem): Boolean {
        if (o is TootStatus) {
            val createdAt = EntityId(o.time_created_at.toString())
            if (createdAt.notDefaultOrConfirming) {
                if (idSet.contains(createdAt)) return true
                idSet.add(createdAt)
            }
        }
        //編集履歴以外のカラムではここを通らない
        return false
    }

    fun filterDuplicateByCreatedAt(src: Collection<TimelineItem>?): ArrayList<TimelineItem> {
        val listNew = ArrayList<TimelineItem>()
        if (src != null) {
            for (o in src) {
                if (isDuplicateByCreatedAt(o)) {
                    log.d("filterDuplicateByCreatedAt: filtered. ${o.cast<TootStatus>()?.time_created_at}")
                    continue
                }
                listNew.add(o)
            }
        }
        return listNew
    }
}
